pgrx/src/datum/borrow.rs

```diff
@@ -0,0 +1,104 @@
+#![deny(unsafe_op_in_unsafe_fn)]
+use super::*;
+use crate::layout::PassBy;
+use core::{ffi, mem, ptr};
+
+/// Types which can be "borrowed from" [`&Datum<'_>`] via simple cast, deref, or slicing
+///
+/// # Safety
+/// Despite its pleasant-sounding name, this implements a fairly low-level detail.
+/// It exists to allow other code to use that nice-sounding BorrowDatum bound.
+/// Outside of the pgrx library, it is probably incorrect to call and rely on this:
+/// instead use the convenience functions available in `pgrx::datum`.
+// TODO: implement those.
+///
+/// Its behavior is trusted for ABI details, and it should not be implemented if any doubt
+/// exists of whether the type would be suitable for passing via Postgres.
+pub unsafe trait BorrowDatum {
+    /// The "native" passing convention for this type.
+    ///
+    /// - `PassBy::Value` implies [`mem::size_of<T>()`][size_of] <= [`mem::size_of::<Datum>()`][Datum].
+    /// - `PassBy::Ref` means the pointee will occupy at least 1 byte for variable-sized types.
+    ///
+    /// Note that this means a zero-sized type is inappropriate for `BorrowDatum`.
+    const PASS: PassBy;
+
+    /// Cast a pointer to this blob of bytes to a pointer to this type.
+    ///
+    /// This is not a simple `ptr.cast()` because it may be *unsizing*, which may require
+    /// reading varlena headers. For all fixed-size types, `ptr.cast()` should be correct.
+    ///
+    /// # Safety
+    /// - This must be correctly invoked for the pointee type, as it may deref and read one or more
+    ///   bytes in its implementation in order to read the inline metadata and unsize the type.
+    ///
+    /// ## For Implementers
+    /// Reading the **first** byte pointed to is permitted if `T::PASS = PassBy::Ref`, assuming you
+    /// are implementing a varlena type. As the other dynamic length type, CStr also does this.
+    /// This function
+    /// - must NOT mutate the pointee
+    /// - must point to the entire datum's length (`size_of_val` must not lose bytes)
+    ///
+    /// Do not attempt to handle pass-by-value versus pass-by-ref in this fn's body!
+    /// A caller may be in a context where all types are handled by-reference, for instance.
+    unsafe fn point_from(ptr: ptr::NonNull<u8>) -> ptr::NonNull<Self>;
+
+    /// Cast a pointer to aligned varlena headers to this type
+    ///
+    /// This version allows you to assume the pointer is aligned to, and readable for, 4 bytes.
+    /// This optimization is not required. When in doubt, avoid implementing it, and rely on your
+    /// `point_from` implementation alone.
+    ///
+    /// # Safety
+    /// - This must be correctly invoked for the pointee type, as it may deref.
+    /// - This must be 4-byte aligned!
+    unsafe fn point_from_align4(ptr: ptr::NonNull<u32>) -> ptr::NonNull<Self> {
+        debug_assert!(ptr.is_aligned());
+        unsafe { <Self as BorrowDatum>::point_from(ptr.cast()) }
```

Comment:
Why does that line use a qualified method call and a few lines below it isn't the case? And which one is actually better in general/for `std` contributions?

Reply:
quantity of paranoia I am subjectively experiencing at the time, mostly

Reply:
@YohDeadfall The main reason I use the fully-qualified form a lot is that this crate has a lot of macros, and in macros, you have to be wary of this: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=7c81475b351b1ee39cc8cc6be3e330fa

There's some clarity benefit sometimes too, but